gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaDeserializerExtractor.java

```diff
@@ -0,0 +1,193 @@
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.source.extractor.extract.kafka;
+
+import java.io.IOException;
+import java.util.Properties;
+
+import org.apache.commons.lang3.reflect.ConstructorUtils;
+
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.Deserializer;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+import io.confluent.kafka.serializers.KafkaAvroDeserializer;
+import io.confluent.kafka.serializers.KafkaJsonDeserializer;
+
+import kafka.message.MessageAndOffset;
+import kafka.utils.VerifiableProperties;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+import gobblin.configuration.WorkUnitState;
+import gobblin.metrics.kafka.KafkaSchemaRegistry;
+import gobblin.metrics.kafka.SchemaRegistryException;
+import gobblin.util.PropertiesUtils;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+
+/**
+ * <p>
+ *   Extension of {@link KafkaExtractor} that wraps Kafka's {@link Deserializer} API. Kafka's {@link Deserializer} provides
+ *   a generic way of converting Kafka {@link kafka.message.Message}s to {@link Object}. Typically, a {@link Deserializer}
+ *   will be used along with a {@link org.apache.kafka.common.serialization.Serializer} which is responsible for converting
+ *   an {@link Object} to a Kafka {@link kafka.message.Message}. These APIs are useful for reading and writing to Kafka,
+ *   since Kafka is primarily a byte oriented system.
+ * </p>
+ *
+ * <p>
+ *   This class wraps the {@link Deserializer} API allowing any existing classes that implement the {@link Deserializer}
+ *   API to integrate with seamlessly with Gobblin. The deserializer can be specified in the following ways:
+ *
+ *   <ul>
+ *     <li>{@link #KAFKA_DESERIALIZER_TYPE} can be used to specify a pre-defined enum from {@link Deserializers}</li>
+ *     <li>{@link #KAFKA_DESERIALIZER_CLASS} can be used to specify the full-qualified name of a {@link Class} that
+ *     defines the {@link Deserializer} interface. If this property is used a {@link KafkaSchemaRegistry} must also
+ *     be specified using the {@link KafkaSchemaRegistry#KAFKA_SCHEMA_REGISTRY_CLASS} config key</li>
+ *   </ul>
+ * </p>
+ */
+@Getter(AccessLevel.PACKAGE)
+public class KafkaDeserializerExtractor extends KafkaExtractor<Object, Object> {
+
+  public static final String KAFKA_DESERIALIZER_TYPE = "kafka.deserializer.type";
+  public static final String KAFKA_DESERIALIZER_CLASS = "kafka.deserializer.class";
```

Comment:
Using two keys can be confusing, it's not clear which one takes precedence.
Can you combine these two? You can make use of `Enums.getIfPresent`